util/hash_test.cc

```diff
@@ -598,6 +613,22 @@ static void test_BitOps() {
     // If we could directly use arithmetic:
     // T vm1 = static_cast<T>(v - 1);
 
+    // BottomNBits
+    {
+      // An essentially full length value
+      T x = everyOtherBit;
+      if (i > 2) {
+        // Make it slightly irregular
+        x = x ^ (T{1} << (i / 2));
+      }
+      auto a = BottomNBits(x, i);
+      auto b = BottomNBits(~x, i);
+      EXPECT_EQ(x | a, x);
+      EXPECT_EQ(a | b, vm1);
```

Comment:
I got confused where this `vm1` gets assigned the `v-1` value to have all last i bits set to 1.

Reply:
It's at the bottom of the loop: `vm1 = (vm1 << 1) | 1;`  And like the earlier comment says, it's done this way because we don't always have arithmetic operators on the 128-bit type available.